### PR TITLE
[SWUI-27] Restructure docs and add SwiftUI.

### PR DESCRIPTION
**Pull request by jeroenleenarts**
### 🔗 Issue Link
SWUI-27

### 🎯 Goal

Restructure the UIKit documentation to have a better grouping of topics to make the docs easier to grasp for endusers.

### 🛠 Implementation

Mostly moving doc markdown a bit to get a better structure.

### 🧪 Testing

Docusaurus does not signal any errors er warnings.

### ☑️ Checklist

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] Changelog is updated with client-facing changes
- [x] New code is covered by unit tests
- [x] Affected documentation updated (docusaurus, tutorial, CMS (task created)

### CODE REVIEW

**Review by petyosi — DISMISSED** [review]

File: docusaurus/docs/iOS/uikit/uikit-overview.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
+---
+title: UIKit Overview
+slug: /uikit
+---
+
+Building on top of the the Stream Chat API, the Stream Chat iOS component libraries include everything you need to build feature-rich and high-functioning chat user experiences out of the box.
+
+We have a component library available for both UIKit. The library includes an extensive set of performant and customizable UI components which allow you to get started quickly with little to no plumbing required. The libraries supports:
```
Comment (nuno-vieira) [inline docusaurus/docs/iOS/uikit/uikit-overview.md:8]: Don't understand the "both"? 🤔 Or it was "both UIKit and SwiftUI" ? 
```suggestion
We have a component library available for UIKit. The library includes an extensive set of performant and customizable UI components which allow you to get started quickly with little to no plumbing required. The libraries support:
```

File: docusaurus/docs/iOS/uikit/uikit-overview.md [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+- Reactions
+- Threads and quoted replies
+- Text input commands (ex: Giphy and @mentions)
+- Image and file uploads
+- Video playback
+- Read state and typing indicators
+- Channel and message lists
+- Push (APN or Firebase)
+- Offline storage
+- OSX
```
Comment (nuno-vieira) [inline docusaurus/docs/iOS/uikit/uikit-overview.md:20]: These are actually supported by the Low Level Client, not the UIKit SDK

File: docusaurus/docs/iOS/uikit/uikit-overview.md [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+#### StreamChatUI
+
+- [Nuke](https://github.com/kean/Nuke) for loading images  
+- [SwiftyGif](https://github.com/kirualex/SwiftyGif) for high performance GIF rendering
+- StreamChat the low-level client to Stream Chat API
+
+#### StreamChat
+
+- [Starscream](https://github.com/daltoniam/Starscream) to handle WebSocket connections
```
Comment (nuno-vieira) [inline docusaurus/docs/iOS/uikit/uikit-overview.md:34]: This is duplicated. Shouldn't we remove this section from the Overview.md and only put this is here on the UIKit version?

File: docusaurus/docs/iOS/basics/overview.md [hunk 4]
```diff
@@ -14,8 +14,6 @@ Building on top of the the Stream Chat API, the Stream Chat iOS component librar
 - Read state and typing indicators
 - Channel and message lists
 - Push (APN or Firebase)
-- Offline storage
-- OSX
```
Comment (nuno-vieira) [inline docusaurus/docs/iOS/basics/overview.md:18]: We should not remove these from the General Overview, only the UIKit Overview 👍

File: docusaurus/docs/iOS/uikit/uikit-overview.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
+---
+title: Overview
+slug: /uikit
+---
+
+We have a component library available for UIKit. The library includes an extensive set of performant and customizable UI components which allow you to get started quickly with little to no plumbing required. The libraries support:
```
Comment (nuno-vieira) [inline docusaurus/docs/iOS/uikit/uikit-overview.md:6]: We are talking about a single library here, not plural
```suggestion
We have a component library available for UIKit. The library includes an extensive set of performant and customizable UI components which allow you to get started quickly with little to no plumbing required. The library support:
```

**Review by adamrushy — APPROVED** [review]
LGTM 👍🏼

File: docusaurus/docs/iOS/basics/overview.md (hunk 4 above)
Comment (nuno-vieira) [reply docusaurus/docs/iOS/basics/overview.md:18]: @jeroenleenarts Can you please put back these 2 options on the "Overview.md" ? These should still be on the general overview

**Review by nuno-vieira — CHANGES_REQUESTED** [review]
Requesting changes, since from the overview of the SDK we should not delete that it has support for Offline and OSX, other then that LGTM 👍

**Review by nuno-vieira — APPROVED** [review]

**Comment by jeroenleenarts** [comment]
Squashed and cleaned up a bunch of the commits into two distinct actions. Restructuring UIKit and adding SwiftUI.

**Review by adamrushy — DISMISSED** [review]

File: docusaurus/docs/iOS/basics/overview.md [hunk 6]
```diff
@@ -38,91 +41,18 @@ This SDK tries to keep the list of external dependencies to a minimum, these are
...
-Add entry to your `Cartfile`:
-
-```
-github "GetStream/stream-chat-swift" ~> 4.0.0
-```
-
-and run:
+- [Nuke](https://github.com/kean/Nuke) for loading images  
+- [NukeUI](https://github.com/kean/Nuke) for loading images  
+- [SwiftyGif](https://github.com/kirualex/SwiftyGif) for high performance GIF rendering
```
Comment (nuno-vieira) [inline docusaurus/docs/iOS/basics/overview.md:50]: Didn't we use Gifu for gifs? 🤔 @martinmitrevski

Comment (martinmitrevski) [reply docusaurus/docs/iOS/basics/overview.md:50]: Good catch! We don't use SwiftyGif, we should remove it from here (I'll remove it from the Package file as well on the other repo).

File: docusaurus/docs/iOS/swiftui/components/attachments.md [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+
+## Message Attachments Overview
+
+The SwiftUI SDK supports the following types of attachments:
+
+- images
+- giphys
+- videos
+- links
+- files
```
Comment (nuno-vieira) [inline docusaurus/docs/iOS/swiftui/components/attachments.md:13]: Maybe?
```suggestion
- Images
- Giphys
- Videos
- Links
- Files
```

File: docusaurus/docs/iOS/swiftui/components/attachments.md [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+        message: message,
+        width: availableWidth,
+        isFirst: isFirst
+    )
+}
+```
+
+Finally, you need to inject the `CustomFactory` into our view hierarchy if you haven't done it already.
+
+Note, if you want to support several different custom views, you will need to do if-else or switch logic in both the message type resolver and the custom view, which will act as container for several views.
```
Comment (nuno-vieira) [inline docusaurus/docs/iOS/swiftui/components/attachments.md:161]: One thing I think is missing from these docs, are some images showing before/after just like we have on the UIKit docs. Without visualising the changes, the docs get pretty boring and easily the reader loses the attention

**Review by nuno-vieira — APPROVED** [review]
LGTM overall, the only thing missing IMO is that we lack images to showcase the customizations. But I guess we can Improve this later 👍

File: docusaurus/docs/iOS/basics/overview.md [hunk 9]
```diff
@@ -47,7 +47,7 @@ This SDK tries to keep the list of external dependencies to a minimum, these are
 
 - [Nuke](https://github.com/kean/Nuke) for loading images  
 - [NukeUI](https://github.com/kean/Nuke) for loading images  
-- [SwiftyGif](https://github.com/kirualex/SwiftyGif) for high performance GIF rendering
+- [Gifu](hhttps://github.com/kaishin/Gifu) for high performance GIF rendering
 - StreamChat the low-level client to Stream Chat API
```
Comment (nuno-vieira) [inline docusaurus/docs/iOS/basics/overview.md]: For UIKit we still use SwiftyGif, please revert this change

**Review by nuno-vieira — CHANGES_REQUESTED** [review]

File: docusaurus/docs/iOS/swiftui/components/helper-views.md [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+Here's an example on how to do that.
+
+```swift
+public func makeChannelListItem(
+    currentChannelId: Binding<String?>,
+    channel: ChatChannel,
+    channelName: String,
+    avatar: UIImage,
+    onlineIndicatorShown: Bool,
+    disabled: Bool,
```
Comment (nuno-vieira) [inline docusaurus/docs/iOS/swiftui/components/helper-views.md:64]: Probably not important for now on the documentation, but we should avoid naming booleans with negation and prefer affirmation. And also, in this case, it is missing the `is` prefix. Ideally this boolean would be `isEnabled`

Comment (martinmitrevski) [reply docusaurus/docs/iOS/swiftui/components/helper-views.md:64]: `disabled` is more in line with SwiftUI's disabled modifier, that actually expects this value (https://developer.apple.com/documentation/swiftui/menu/disabled(_:)). But yes, we can discuss this further - it can be either way.

File: docusaurus/docs/iOS/swiftui/components/message-composer.md [hunk 11]
```diff
@@ -0,0 +1,395 @@
...
+
+For example, let's say we want to add additional contacts picker, which will allow us to send contacts via the chat. We also want to keep the existing functionalities, therefore we will explore ways to customize the existing component.
+
+There are few things we need to do in order to accomplish this: 
+- introduce new type of payload for contacts
+- create new contact attachment type component
+- enable the new contact component to be selectable from the attachment types picker
+- implement the UI for previewing the selected item (contact) in the message composer
+- update the message resolving logic of the message list to include the contact payload
+- provide UI for the contacts attachment in the message list
```
Comment (nuno-vieira) [inline docusaurus/docs/iOS/swiftui/components/message-composer.md:42]:
```suggestion
- Introduce a new type of payload for contacts
- Create a new contact attachment type component
- Enable the new contact component to be selectable from the attachment types picker
- Implement the UI for previewing the selected item (contact) in the message composer
- Update the message resolving logic of the message list to include the contact payload
- Provide UI for the contacts attachment in the message list
```

File: docusaurus/docs/iOS/swiftui/components/attachments.md (hunk 8 above)
Comment (martinmitrevski) [reply docusaurus/docs/iOS/swiftui/components/attachments.md:161]: That's good point. @jeroenleenarts, in the next version we can put the ones from your tutorial.

File: docusaurus/docs/iOS/swiftui/components/message-composer.md [hunk 12]
```diff
@@ -0,0 +1,395 @@
...
+
+struct ContactAttachmentPayload: AttachmentPayload {
+    static let type: AttachmentType = .contact
+
+    let name: String
+    let phoneNumber: String
+}
+```
+
+Since we will go through the attachments in a scrollable list, we also need to conform the payload to the `Identifiable` protocol. For an id, it's enough to use a key that's combination of the name and the phone number.
```
Comment (nuno-vieira) [inline docusaurus/docs/iOS/swiftui/components/message-composer.md:63]:
```suggestion
Since we will go through the attachments in a scrollable list, we also need to conform the payload to the `Identifiable` protocol. For an id, it's enough to use a key that's a combination of the name and the phone number.
```

File: docusaurus/docs/iOS/swiftui/components/helper-views.md (hunk 10 above)
Comment (nuno-vieira) [reply docusaurus/docs/iOS/swiftui/components/helper-views.md:64]: Really weird, was not expecting that 🤔 If that is the case, and is consistent with SwiftUI, then it is fine for me 🙂

File: docusaurus/docs/iOS/swiftui/components/channel-list-header.md [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+
+In most cases, you will need to customize the channel list header to fit in with the rest of your app. The SwiftUI SDK provides several customization options, from minor tweaks to ultimately providing your own header.
+
+The most simple change you can do is to change the title of the header, while keeping the same look and feel of it. In order to do this, simply pass the title in the initializer of the `ChatChannelListScreen` or `ChatChannelListView` component.
+
+```swift
+var body: some Scene {
+    WindowGroup {
+		ChatChannelListScreen(title: "Custom title")
+    }
```
Comment (nuno-vieira) [inline docusaurus/docs/iOS/swiftui/components/channel-list-header.md:15]:
```suggestion
    WindowGroup {
        ChatChannelListScreen(title: "Custom title")
    }
```

File: docusaurus/docs/iOS/swiftui/components/message-composer.md [hunk 14]
```diff
@@ -0,0 +1,395 @@
...
+        onCustomAttachmentTap: @escaping (CustomAttachment) -> Void
+) -> some View {
+    CustomContactAttachmentComposerPreview(
+        addedCustomAttachments: addedCustomAttachments,
+        onCustomAttachmentTap: onCustomAttachmentTap
+    )    
+}
+```
+
+In this view, we will just re-use the `CustomContactAttachmentPreview` we've created above (only without the checkmark functionality). In addition, we are adding the `DiscardAttachmentButton` from the SDK, to allow the possibility to remove the attachment from the composer's input view. You can provide your own version of this button, if needed.
```
Comment (nuno-vieira) [inline docusaurus/docs/iOS/swiftui/components/message-composer.md:291]:
```suggestion
In this view, we will just re-use the `CustomContactAttachmentPreview` we've created above (only without the checkmark functionality). In addition, we are adding the `DiscardAttachmentButton` from the SDK, to allow the possibility to remove the attachment from the composer's input view. You can provide your own version of this button if needed. 
```

File: docusaurus/docs/iOS/swiftui/components/message-reactions.md [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+public func makeMessageReactionView(
+        message: ChatMessage
+) -> some View {
+    CustomReactionsContainer(message: message)
+}
+```
+
+## Customizing the Reactions Overlay View
+
+The reactions overlay view (shown on long press of a message), also provides access to the message actions. If you want to replace / extend the default message actions, you can do so via the `supportedMessageActions` method in the `ViewFactory`. As an inspiration, here's a glimpse on how the default message actions are configured.
```
Comment (nuno-vieira) [inline docusaurus/docs/iOS/swiftui/components/message-reactions.md:39]:
```suggestion
The reactions overlay view (shown on long-press of a message), also provides access to the message actions. If you want to replace/extend the default message actions, you can do so via the `supportedMessageActions` method in the `ViewFactory`. As an inspiration, here's a glimpse of how the default message actions are configured.
```

File: docusaurus/docs/iOS/swiftui/components/message-reactions.md [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+
+## Customizing the Message Reactions View
+
+The simplest way to customize the message reactions view is to replace its reaction icons. Those are available under the `availableReactions` property in the Images class, which is part of the Appearance class in the StreamChat object. The `availableReactions` property is a dictionary, which contains mappings between `MessageReactionType` and its corresponding `ChatMessageReactionAppearanceType`, which consists of small and large icon for a reaction. If you change these properties, make sure to inject the updated `Images` class in the StreamChat object.
+
+```swift
+let customReactions = [.init(rawValue: "custom"): ChatMessageReactionAppearance(
+                        smallIcon: reactionCustomSmall,
+                        largeIcon: reactionCustomBig
+                      )]
```
Comment (nuno-vieira) [inline docusaurus/docs/iOS/swiftui/components/message-reactions.md:17]: For future improvement we can use StringByLiteral to improve the API here
```suggestion
let customReactions = ["custom": ChatMessageReactionAppearance(
                        smallIcon: reactionCustomSmall,
                        largeIcon: reactionCustomBig
                      )]
```

**Review by nuno-vieira — APPROVED** [review]
LGTM ✅ Only small nits
